server/README.md

```diff
@@ -0,0 +1,236 @@
+
+A very simple way to ensure causality and concurrency on certificates would be:
+
+- Have a single big lock in the server that must be taken before any attempt to
+  add a certificate (handles concurrency).
+- Ensure the new certificate's timestamp is newer than any existing certificate
+  (handles causality).
+
+However this is pretty wasteful given most certificates are unrelated (e.g. certif A
+add a new user, certif B does a key rotation in a realm). So instead we reduce the
+size of the lock by introducing the concept of topic.
+
+Each type of certificate is part of one of the following topics:
+
+- `common`
+- `sequester`, which depends on `common`
+- `realm`, which depends on `common` and uses the realm ID as discriminant
+- `shamir_recovery`, which depends on `common` and use the ID of the user to recover as discriminant
+
+So the actual rule used in the server is:
+
+- Have lock in the server for each couple topic + discriminant
+- Take the lock for the topic + discriminant couple the certificate we want to add
+  belongs to AND to other topics our topic may depend on (handles concurrency).
+- Ensure the new certificate's timestamp is newer than any existing certificate
+  in the topics we have locked (handles causality).
+
+> Note: Also organizations are totally isolated from each other on the server, so
+> in a way the organization ID is an aditional discriminant on each certificate topic.
+
+### Blocks
+
+Blocks are encrypted piece of data that belong to a given realm and cannot
+be modified.
+
+This makes concurrency & causality pretty simple:
+
+- Each block is considered independent, hence there is no need for lock dedicated to blocks.
+- A block is identified by a client-provided UUID, then there is no risk for double-creation
+  under concurrency (a trivial unicity check on insertion prevents this).
+- Block creation date must be more recent than the last certificate in `common` topic and
+  in the `realm` topic corresponding to the block's realm.
+
+> Notes:
+>
+> - It could be tempting to only lock the `realm` topic (given it's this topic that contains
+>   the certificate describing the role of a user in the realm). However a revoked user is also
+>   forbidden from adding data, hence making it mandatory to also lock the `common` topic here.
+> - Uploading blocks (and vlobs) are the most common task on the server, so we may want to
+>   uptomise it in the future to avoid having to lock two topics.
+>   A possible solution would be to have an optimistic locking approach: first read the
+>   last timestamp in the topics, then do the checks, then on the final data insertion check
+>   again the timestamp haven't changed (and if they have we re-do everything from the start).
+
+### Vlobs
+
+Vlobs are encrypted metadata that belong to a given realm. Unlike block they can be
+modified and hence have a version field that must increase linearly.
+
+Concurrency & causality is handle is a similar fashion than for blocks.
```

Comment:
Really? Vlobs require a database write, but blocks require bother a database write and upload to the object storage. I would find very surprising if both used the same approach in the end.

Reply:
You can think of both vlob and block upload as:

1) fetch info from database
2) do stuff in Python
3) store the vlob/block in the database

In the case of vlob, step 2 is only about checking access rights and vlob existence. On the other hand block also includes the write to the object storage.

However this is irrelevant from the database point of view: in both case step 3 is only allowed if data in step 1 hasn't changed.

Obviously this approach means we can end up with data in the  object storage while the SQL transaction got cancelled.
However this is acceptable:
- Given the block ID is determined client-side, re-trying to upload the block will lead to another write in the object storage at the same location.
- The object storage follows an eventual consistency strategy, hence the new block upload is guaranteed to overwrite the previous one. The only issue would be if the client upload a different block with the same ID between each try, in this case eventual consistency means it is possible for a short period of time to get a previous value when querying the block (but this is purely theoretical since there is no reason for the client to do that !).